authorization/storage_authorization.go

```diff
@@ -226,41 +345,120 @@ func (s *Store) forEachAuthorization(ctx context.Context, tx kv.Tx, pred kv.Curs
 	return nil
 }
 
-// UpdateAuthorization updates the status and description only of an authorization
-func (s *Store) UpdateAuthorization(ctx context.Context, tx kv.Tx, id platform.ID, a *influxdb.Authorization) (auth *influxdb.Authorization, retErr error) {
-	defer func() {
-		retErr = errors.ErrInternalServiceError(retErr, errors.WithErrorOp(influxdb.OpUpdateAuthorization))
-	}()
-	v, err := encodeAuthorization(a)
+// commitAuthorization performs pre-commit checks and updates to an authorization record, commits it,
+// and makes sure indices point to it. It does not delete any indices. The updated authorization is
+// returned on success.
+func (s *Store) commitAuthorization(ctx context.Context, tx kv.Tx, a *influxdb.Authorization) error {
+	if err := s.verifyTokensMatch(a); err != nil {
+		return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInvalid))
+	}
+
+	if err := s.hashToken(a); err != nil {
+		return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInternal))
+	}
+
+	v, err := s.encodeAuthorization(a)
 	if err != nil {
-		return nil, errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInvalid))
+		return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInvalid))
 	}
 
 	encodedID, err := a.ID.Encode()
 	if err != nil {
-		return nil, errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.ENotFound))
+		return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.ENotFound))
 	}
 
-	idx, err := authIndexBucket(tx)
-	if err != nil {
-		return nil, err
+	if !s.useHashedTokens && a.Token != "" {
+		idx, err := authIndexBucket(tx)
+		if err != nil {
+			return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInternal))
+		}
+
+		if err := idx.Put(authIndexKey(a.Token), encodedID); err != nil {
+			return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInternal))
+		}
 	}
 
-	if err := idx.Put(authIndexKey(a.Token), encodedID); err != nil {
-		return nil, err
+	if a.HashedToken != "" {
+		idx, err := hashedAuthIndexBucket(tx)
+		// Don't ignore a missing index here, we want an error.
+		if err != nil {
+			return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInternal))
+		}
+
+		if err := idx.Put(hashedAuthIndexKey(a.HashedToken), encodedID); err != nil {
+			return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInternal))
+		}
 	}
 
 	b, err := tx.Bucket(authBucket)
 	if err != nil {
-		return nil, err
+		return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInternal))
 	}
 
 	if err := b.Put(encodedID, v); err != nil {
+		return errors.ErrInternalServiceError(err, errors.WithErrorCode(errors.EInternal))
+	}
+
+	return nil
+}
+
+// deleteIndices removes indices for the given token and hashedToken.
+func (s *Store) deleteIndices(ctx context.Context, tx kv.Tx, token, hashedToken string) error {
```

Comment:
I'm wondering if this should be more robust. Even if we can't get one bucket, remove the (hashed) token from the other, and if we can't remove one token, still try to remove the other.  Would that leave us in a better state for recovery, or not?

Reply:
If error will get returned up the call tree and cause the BoltDB transaction to be rolled back, so we won't leave BoltDB in an inconsistent state.